src/fsharp/vs/IncrementalBuild.fs

```diff
@@ -1641,17 +1655,56 @@ type IncrementalBuilder(tcGlobals,frameworkTcImports, nonFrameworkAssemblyInputs
     let MaxTimeStampInDependencies cache (ctok: CompilationThreadToken) (output:INode) = 
         IncrementalBuild.MaxTimeStampInDependencies cache ctok output.Name partialBuild 
 
-    member this.IncrementUsageCount() = 
-        assertNotDisposed() 
-        System.Threading.Interlocked.Increment(&referenceCount) |> ignore
-        { new System.IDisposable with member x.Dispose() = this.DecrementUsageCount() }
+    /// Try to increment the usage count. If it returns None, then the object has been disposed by someone else already and may no longer be used.
+    member this.TryIncrementUsageCount() =
+        // Why is this so complex?
```

Comment:
Could you abstract all this out to a helper somewhere, e.g. in illib.fs please?   It shouldn't be hidden away here.  We might well need to reuse it.

Still doesn't seem it should be this complex, but I'll believe you :)